Compile rustc crates with the initial-exec TLS model

joshtriplett opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/78201)*
<!-- homu-ignore:end -->

This should produce more efficient code, with fewer calls to
__tls_get_addr. The tradeoff is that libraries using it won't work with
dlopen, but that shouldn't be a problem for rustc's internal libraries.

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

joshtriplett commented:
cc @eddyb

joshtriplett commented:
Trying this out to see if it provides a performance improvement.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 328401119c35ff10758630aafe5eb7a04925c399 with merge a612e191b94bc6e23963dbadbab26b6486a51dc3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"328401119c35ff10758630aafe5eb7a04925c399","merge_sha":"a612e191b94bc6e23963dbadbab26b6486a51dc3"} -->

eddyb commented on src/bootstrap/bin/rustc.rs:
```diff
@@ -104,6 +104,13 @@ fn main() {
+        // Compile rustc crates with the more efficient initial-exec TLS model.
+        // This doesn't work with `dlopen`, so we can't use it by default in
+        // general, but we can use it for rustc's own internal libraries.
+        if crate_name.map_or(false, |n| n.starts_with("rustc_") || n.starts_with("rustc-")) {
+            cmd.arg("-Z").arg("tls-model=initial-exec");
+        }
```
I wonder if we should also check that the path includes `compiler/rustc_` because there's some stuff on crates.io that also starts with `rustc_` or `rustc-` and it might not be strictly a `rustc_driver` dependency.

joshtriplett replied:
We need a much better way to handle this, yes; this was just the simplest way to test the impact of this change.

Mark-Simulacrum replied:
Best bet is probably to modify builder.rs to set RUSTC_IS_COMPILING_THE_COMPILER for cargo and then read that here.

eddyb commented:
This may be silly, but, if we want to exactly track which crates are using TLS, so that we can list them out in the wrapper, we could:
* have an allow-by-default lint against using `#[thread_local]` (including from `thread_local!`)
* pass `-Fthread-local` from the `bootstrap/bin/rustc.rs` wrapper by default for all crates
  * specifically using "forbid" means `#[allow(thread_local)]` wouldn't work so it couldn't be accidentally used
* for an explicit list of crates, we would omit `-Fthread-local`
  * the subset of that which ends up linked into `librustc_driver-*.so`, also get `-Ztls-model=initial-exec`

This could get confusing so we might want to make it an internal `rustc::untracked_thread_local` lint that has a custom message telling you to what list to add the crate etc.

joshtriplett commented:
@eddyb If we're going to go *that* far, I would rather just see this mechanism fixed to make the TLS model depend on the top-level artifact(s), such as "if you're building a binary use local-exec, if you're building a cdylib use global-dynamic for pub and local-dynamic for private".

eddyb commented:
> If we're going to go that far, I would rather just see this mechanism fixed to make the TLS model depend on the top-level artifact(s)

Oh yeah I forgot I wanted that too (should be possible, based on my experiments an `rlib` crate is undecided between static and dynamic TLS until it's linked into a `bin` crate, or a `dylib` crate, respectively) and went down the rabbit hole of thinking about tracking which crates define `#[thread_local]` `static`s.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1289094690), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37351)
Build commit: a612e191b94bc6e23963dbadbab26b6486a51dc3 (`a612e191b94bc6e23963dbadbab26b6486a51dc3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37351","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1289094690"},"merge_sha":"a612e191b94bc6e23963dbadbab26b6486a51dc3"} -->

rust-timer commented:
Queued a612e191b94bc6e23963dbadbab26b6486a51dc3 with parent 1d2726726f8f3128e98191e4c6cb94bd76d0ddd4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1d2726726f8f3128e98191e4c6cb94bd76d0ddd4&end=a612e191b94bc6e23963dbadbab26b6486a51dc3).

rust-timer commented:
Finished benchmarking try commit (a612e191b94bc6e23963dbadbab26b6486a51dc3): [comparison url](https://perf.rust-lang.org/compare.html?start=1d2726726f8f3128e98191e4c6cb94bd76d0ddd4&end=a612e191b94bc6e23963dbadbab26b6486a51dc3).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

joshtriplett commented:
Looks like this is definitely worth doing, and we just need to figure out the best way to capture the set of crates we can safely do it for.

Ideally, we'd apply this to every crate that isn't a proc macro or compiled into a dynamically linkable version of `std`. How can we most easily capture that set here?

bjorn3 commented:
If you tell the linker about the storage model instead of LLVM it should rewrite all TLS accesses to use that storage model I think. You could do that using a build.rs for rustc_driver. That will apply it to exactly the dylib that it should be applied to.

Mark-Simulacrum commented:
Hm, so when you ask "Ideally, we'd apply this to every crate that isn't a proc macro or compiled into a dynamically linkable version of std. How can we most easily capture that set here?" -- do you mean to do that as an end-user visible change (rather than just to crates compiled by rustbuild? It seems to me that rustc shouldn't really be special here in some sense. That's probably not quite true because we have the somewhat odd case of compiling to dylibs that aren't intended to be dlopen'd (rather just linked directly), but I'm not sure that's important.

I think the best thing is probably to start out with just a change to how we compile rustc which should be somewhat easier to land and see impacts of though. To do so we probably want to set the -Z flag via rustflags if possible, something like this diff. This will set it for pretty much everything except std workspace and crates that don't get rustflags (e.g., proc macros and other build deps). Note that this includes e.g. cargo and clippy too, but I think that's the right thing?

```diff
diff --git a/src/bootstrap/builder.rs b/src/bootstrap/builder.rs
index 707c1ff3efa..0f6a74b01ce 100644
--- a/src/bootstrap/builder.rs
+++ b/src/bootstrap/builder.rs
@@ -1114,6 +1114,10 @@ impl<'a> Builder<'a> {
             }
         }
 
+        if mode != Mode::Std {
+            rustflags.arg("-Ztls-model=initial-exec");
+        }
+
         if self.config.incremental {
             cargo.env("CARGO_INCREMENTAL", "1");
         } else {
```

joshtriplett committed:
Compile tools and internal libraries with the initial-exec TLS model

This should produce more efficient code, with fewer calls to
__tls_get_addr. The tradeoff is that libraries using it won't work with
dlopen, but that shouldn't be a problem for tools or for our own
internal libraries.

Co-authored-by: Mark Rousskov <mark.simulacrum@gmail.com>

joshtriplett commented:
@Mark-Simulacrum Yeah, that looks perfect. That's exactly the set of crates that should work with this.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 15068d492e48e3aefc563baa5aa9bd6ee9223a80 with merge e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37...
<!-- homu: {"type":"TryBuildStarted","head_sha":"15068d492e48e3aefc563baa5aa9bd6ee9223a80","merge_sha":"e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1307167038)
Build commit: e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37 (`e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1307167038"},"merge_sha":"e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37"} -->

rust-timer commented:
Queued e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37 with parent 16e9ed0b1c58e0327eb37eb6f70e9b9ef1844591, future [comparison URL](https://perf.rust-lang.org/compare.html?start=16e9ed0b1c58e0327eb37eb6f70e9b9ef1844591&end=e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37).

rust-timer commented:
Finished benchmarking try commit (e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37): [comparison url](https://perf.rust-lang.org/compare.html?start=16e9ed0b1c58e0327eb37eb6f70e9b9ef1844591&end=e33e3c8a267474be0fabf07c0eb7b6e3f7e00e37).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

petrochenkov commented:
@bjorn3 
>If you tell the linker about the storage model instead of LLVM it should rewrite all TLS accesses to use that storage model I think.

How do you do that? I can't find any options on [`ld` doc page](https://sourceware.org/binutils/docs/ld/Options.html) that look relevant, at least by searching for thread/tls/etc.

bjorn3 commented:
I found `-z nodlopen`, but it doesn't seem to affect tls storage model. By the way instead of `mode != Mode::Std` I think it should be `mode == Mode::Rustc`, so other things don't accidentally get the initial-exec TLS storage model.

joshtriplett commented:
@bjorn3 I looked at the other modes, and I think all of them other than std *should* use this.

bjorn3 commented:
Once https://github.com/rust-lang/rust/pull/77975 lands `Mode::Codegen` shouldn't use this. Also for executables the linker will rewrite it anyway.

joshtriplett commented:
@bjorn3 Does the cranelift backend need to be openable via dlopen? In general, it *seems* like codegen backends should be able to use this, and it'd be important for them to be able to do so because they're a lot of code.

bjorn3 commented:
cg_clif is only loaded using dlopen. Adding it as dependency of rustc would require specifying the path to every rustc dependency instead of loading them from the sysroot. This doesn't work for out of tree development.

joshtriplett commented:
@bjorn3 I thought that was an artifact of cg_clif being out-of-tree, and that it would become less of an issue once it wasn't.

But in any case, I can certainly add `Mode::Codegen` to the list. Will that prevent optimizing the LLVM backend's use of TLS, if any?

bjorn3 commented:
> @bjorn3 I thought that was an artifact of cg_clif being out-of-tree, and that it would become less of an issue once it wasn't.

The integration uses `git subtree`. I will continue developing it out of tree and syncing it once in a while.

> But in any case, I can certainly add Mode::Codegen to the list. Will that prevent optimizing the LLVM backend's use of TLS, if any?

The LLVM backend is compiled as dependency lf rustc_interface and thus uses `Mode::Rustc`.

joshtriplett commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 36747df223fd06ba7198ceab8a5980bbfbec388d with merge c359cffb64d0a55cc9c65af57152354514f70f29...
<!-- homu: {"type":"TryBuildStarted","head_sha":"36747df223fd06ba7198ceab8a5980bbfbec388d","merge_sha":"c359cffb64d0a55cc9c65af57152354514f70f29"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1360908945)
Build commit: c359cffb64d0a55cc9c65af57152354514f70f29 (`c359cffb64d0a55cc9c65af57152354514f70f29`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1360908945"},"merge_sha":"c359cffb64d0a55cc9c65af57152354514f70f29"} -->

rust-timer commented:
Queued c359cffb64d0a55cc9c65af57152354514f70f29 with parent 9d78d1d02761b906038ba4d54c5f3427f920f5fb, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9d78d1d02761b906038ba4d54c5f3427f920f5fb&end=c359cffb64d0a55cc9c65af57152354514f70f29).

rust-timer commented:
Finished benchmarking try commit (c359cffb64d0a55cc9c65af57152354514f70f29): [comparison url](https://perf.rust-lang.org/compare.html?start=9d78d1d02761b906038ba4d54c5f3427f920f5fb&end=c359cffb64d0a55cc9c65af57152354514f70f29).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

nagisa commented on src/bootstrap/builder.rs:
```diff
@@ -1140,6 +1140,14 @@ impl<'a> Builder<'a> {
+        // efficient initial-exec TLS model. This doesn't work with `dlopen`,
+        // so we can't use it by default in general, but we can use it for tools
+        // and our own internal libraries.
+        if mode != Mode::Std && mode != Mode::Codegen {
```
Might there be value in exhaustively matching the modes to ensure that we don't accidentally apply this flag to wrong things in the future, when somebody adds a new mode(s)? I could see it also making sense as a method along the lines of `Mode::is_dlopenable` or something...

joshtriplett replied:
@nagisa I don't think it's worth exhaustively matching, but I do think factoring out a method makes sense. Done.

bjorn3 commented on src/bootstrap/builder.rs:
```diff
@@ -1140,6 +1140,14 @@ impl<'a> Builder<'a> {
+        // efficient initial-exec TLS model. This doesn't work with `dlopen`,
+        // so we can't use it by default in general, but we can use it for tools
+        // and our own internal libraries.
+        if mode.must_support_dlopen() {
```
```suggestion
        if !mode.must_support_dlopen() {
```

joshtriplett replied:
@bjorn3 Good catch, thank you.

Mark-Simulacrum commented:
r=me with @bjorn3's nit fixed

joshtriplett commented:
@bjorn3 Good catch, thank you.

joshtriplett commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 0328e69287b083af4d5d4b49cfc9175e9c82c88e has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 0328e69287b083af4d5d4b49cfc9175e9c82c88e -->
<!-- homu: {"type":"Approved","sha":"0328e69287b083af4d5d4b49cfc9175e9c82c88e","approver":"Mark-Simulacrum"} -->

joshtriplett commented:
Labeling relnotes as a heads-up for anyone who might be doing something especially creative involving dynamic loading of parts of rustc.

bors commented:
:hourglass: Testing commit 0328e69287b083af4d5d4b49cfc9175e9c82c88e with merge 05ed67b103ab3e2a9ddf6fa5cde06050a4213f92...
<!-- homu: {"type":"BuildStarted","head_sha":"0328e69287b083af4d5d4b49cfc9175e9c82c88e","merge_sha":"05ed67b103ab3e2a9ddf6fa5cde06050a4213f92"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1367645285)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1367645285","builder_name":"checks-actions"} -->

m-ou-se commented:
Looks like the spurious error that's been happening randomly on the Apple builders lately.

@bors retry

bors commented:
:hourglass: Testing commit 0328e69287b083af4d5d4b49cfc9175e9c82c88e with merge a257bc7123690e623513a0a7b8819f2401003514...
<!-- homu: {"type":"BuildStarted","head_sha":"0328e69287b083af4d5d4b49cfc9175e9c82c88e","merge_sha":"a257bc7123690e623513a0a7b8819f2401003514"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1371122683)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1371122683","builder_name":"checks-actions"} -->

jyn514 commented:
Looks spurious?

```
------------------------------------------
stderr:
------------------------------------------
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
clang: CommandLine Error: Option 'h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
clang: CommandLine Error: Option 'h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

------------------------------------------
```

Mark-Simulacrum commented:
@bors retry - agreed that this seems spurious

```
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
clang: CommandLine Error: Option 'h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
clang: CommandLine Error: Option 'h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine **options*
```

bors commented:
:hourglass: Testing commit 0328e69287b083af4d5d4b49cfc9175e9c82c88e with merge e012fa10bed4afe82574679760549a76eb053951...
<!-- homu: {"type":"BuildStarted","head_sha":"0328e69287b083af4d5d4b49cfc9175e9c82c88e","merge_sha":"e012fa10bed4afe82574679760549a76eb053951"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1371677197)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1371677197","builder_name":"checks-actions"} -->

mati865 commented:
The same error again on apple.
Maybe it should be enabled only on Linux?

Aaron1011 commented:
This is a known issue which is unrelated to this PR: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/apple-x86_64.20gha.20checks

joshtriplett commented:
@bors retry

bors commented:
:hourglass: Testing commit 0328e69287b083af4d5d4b49cfc9175e9c82c88e with merge 25f6938da459a57b43bdf16ed6bdad3225b2a3ce...
<!-- homu: {"type":"BuildStarted","head_sha":"0328e69287b083af4d5d4b49cfc9175e9c82c88e","merge_sha":"25f6938da459a57b43bdf16ed6bdad3225b2a3ce"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1374427690)
Approved by: Mark-Simulacrum
Pushing 25f6938da459a57b43bdf16ed6bdad3225b2a3ce to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1374427690"},"merge_sha":"25f6938da459a57b43bdf16ed6bdad3225b2a3ce"} -->

glaubitz commented:
This change broke the Rust compiler on 32-bit PowerPC, see: #81334

viferga commented:
Can you revert this? I am embedding `rustc_driver` for a long time in a plugin system and recently I updated the toolchain and you break my system with this.

Optimizing a negligible amount of time by sacrificing dlopen is not something valid to me.

Project: https://github.com/metacall/core/tree/develop/source/loaders/rs_loader
PR with new toolchain: https://github.com/metacall/core/pull/716
Broken CI from this exact issue: https://github.com/metacall/core/actions/runs/25819621618/job/75860817819
Error:

```Error : DynLink error: /root/.rustup/toolchains/nightly-2026-01-15-x86_64-unknown-linux-gnu/lib/librustc_driver-6b3a4a0144cdc0dc.so: cannot allocate memory in static TLS block```

lqd commented:
> recently I updated the toolchain

This was merged 6 years ago, are you commenting on the correct PR instead of a more recent one?

viferga commented:
> > recently I updated the toolchain
> 
> This was merged 6 years ago, are you commenting on the correct PR instead of a more recent one?

This is the first commit introducing this flag which generates the issue. I was using this toolchain before and it was working:
https://github.com/metacall/core/blob/develop/source/loaders/rs_loader/rust/rust-toolchain

This is the issue: https://users.rust-lang.org/t/cannot-allocate-memory-in-static-tls-block/133392/2

lqd commented:
Make sure to do https://users.rust-lang.org/t/cannot-allocate-memory-in-static-tls-block/133392/6

viferga commented:
> Make sure to do https://users.rust-lang.org/t/cannot-allocate-memory-in-static-tls-block/133392/6

Man are you reading my post and the reference I left? I am using dlopen I cannot do that, it's a plugin system.

lqd commented:
The rustc driver is made with the rust executables in mind: it exists to avoid having rustc, clippy, rustfmt etc duplicate all that code. We also use unstable features to build it (LTO on dylibs for example) because we know this limited context and can test all these use-cases. It’s not a public API, and we could reverse its existence in the future.

You can build and distribute your own with the changes you need but that is also suboptimal.

Either way, I suggest bringing this up on zulip rather than on a 6 year old PR that no one would notice, as well as being less rude to the people trying to help you, and will unsubscribe from this thread.

viferga commented:
> The rustc driver is made with the rust executables in mind: it exists to avoid having rustc, clippy, rustfmt etc duplicate all that code. We also use unstable features to build it (LTO on dylibs for example) because we know this limited context and can test all these use-cases. It’s not a public API, and we could reverse its existence in the future.
> 
> You can build and distribute your own with the changes you need but that is also suboptimal.
> 
> Either way, I suggest bringing this up on zulip rather than on a 6 year old PR that no one would notice, as well as being less rude to the people trying to help you, and will unsubscribe from this thread.

I understand, I thought you were gaslighting me.

But how the TLS mode is related to what you said? And also how linking rustc_driver is related to clippy, rustfmt etc?

What do you mean by suboptimal? If you don't change this I won't have any other option than build the toolchain myself and patching the build system.

Where can I find the zulip link?

jyn514 commented:
https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp

mati865 commented:
> And also how linking rustc_driver is related to clippy, rustfmt etc?

`rustc_driver` is a compiler internal library used by rustc, Clippy, Rustfmt and other compiler related binaries. You are free to make it work with your own tooling, but that's not something that we support currently.

viferga commented:
I already know what's clippy etc. I am asking how is this related to the topic.

I will try to talk about this in zulip.

What I'm talking about is that you're distributing in linux this library as a .so. But then with this PR is being forced to be linked like a static library, disabling dlopen on it.

I don't think cutting this functionality by forcing this TLS mode makes sense, and also this breaks my system. And apart from that if this is something you do for performance, I don't see the benchmarks for verifying the optimization.

bjorn3 commented:
> I don't see the benchmarks for verifying the optimization.

It was an up to 7% perf win at the time: https://github.com/rust-lang/rust/pull/78201#issuecomment-722780636 Most optimizations we do nowadays have nowhere near that effect, so losing this perf win would be rather painful. And it is probably even higher nowadays given that we are no longer optimizing as hard to avoid TLS accesses given that this PR makes TLS accesses almost free.

> What I'm talking about is that you're distributing in linux this library as a .so. But then with this PR is being forced to be linked like a static library, disabling dlopen on it.

The main reason we distribute rustc_driver as dylib at all is to reduce the size of the toolchain by allowing the rustc internals to be shared between tools we ship without each executable having to duplicate all this machine code. The only cases where we use dlopen at all are proc macros (which intentionally statically link everything and talk over a C ABI with rustc) and hotplug codegen backends (for which `-Zexport-executable-symbols` when compiling the rustc executable should also work to get access to rustc_driver)

viferga commented:
> > I don't see the benchmarks for verifying the optimization.
> 
> It was an up to 7% perf win at the time: [#78201 (comment)](https://github.com/rust-lang/rust/pull/78201#issuecomment-722780636) Most optimizations we do nowadays have nowhere near that effect, so losing this perf win would be rather painful. And it is probably even higher nowadays given that we are no longer optimizing as hard to avoid TLS accesses given that this PR makes TLS accesses almost free.
> 
> > What I'm talking about is that you're distributing in linux this library as a .so. But then with this PR is being forced to be linked like a static library, disabling dlopen on it.
> 
> The main reason we distribute rustc_driver as dylib at all is to reduce the size of the toolchain by allowing the rustc internals to be shared between tools we ship without each executable having to duplicate all this machine code. The only cases where we use dlopen at all are proc macros (which intentionally statically link everything and talk over a C ABI with rustc) and hotplug codegen backends (for which `-Zexport-executable-symbols` when compiling the rustc executable should also work to get access to rustc_driver)

So basically there's no way you'll revert this or add rustc_driver to the white list?

bjorn3 commented:
> no way you'll revert this

Probably not

> or add rustc_driver to the white list?

The whole reason this improves performance is that rustc_driver and all dependencies are getting compiled with this flag.

viferga commented:
> > no way you'll revert this
> 
> Probably not
> 
> > or add rustc_driver to the white list?
> 
> The whole reason this improves performance is that rustc_driver and all dependencies are getting compiled with this flag.

Ok, thanks. I will fork it then.

But this disables the possibility to embed your compiler.

mati865 commented:
> I already know what's clippy etc. I am asking how is this related to the topic.

> What I'm talking about is that you're distributing in linux this library as a .so. But then with this PR is being forced to be linked like a static library, disabling dlopen on it.

As I said, it's an _internal library_ and meant to be used by the mentioned tools that Rust distributes.

> But this disables the possibility to embed your compiler.

Maybe we can figure another way to solve your problem, but digging up 6-year-old PRs isn't the right place for such discussions.

viferga commented:
> > I already know what's clippy etc. I am asking how is this related to the topic.
> 
> > What I'm talking about is that you're distributing in linux this library as a .so. But then with this PR is being forced to be linked like a static library, disabling dlopen on it.
> 
> As I said, it's an _internal library_ and meant to be used by the mentioned tools that Rust distributes.
> 
> > But this disables the possibility to embed your compiler.
> 
> Maybe we can figure another way to solve your problem, but digging up 6-year-old PRs isn't the right place for such discussions.

What do you propose?

mati865 commented:
Using Zulip or users.rust-lang.org as the first step.

viferga commented:
We solved the issue: https://github.com/metacall/rust-toolchain
